docs/udf.rst

```diff
@@ -0,0 +1,179 @@
+different ways:
+
+- with two int parameters returning a long result.
+- with two long parameters returning a long result.
+- with two double parameters returning a double result.
+
+.. code:: java
+
+    @UdfDescription(name = "multiply", description = "multiplies 2 numbers")
+    public class Multiply {
+
+      @Udf(description = "multiply two ints")
+      public long multiply(final int v1, final int v2) {
+        return v1 * v2;
+      }
+
+      @Udf(description = "multiply two longs")
+      public long multiply(final long v1, final long v2) {
+        return v1 * v2;
+      }
+
+      @Udf(description = "multiply two doubles")
+      public double multiply(final double v1, double v2) {
+        return v1 * v2;
+      }
+
+    }
+
+Supported Types
+---------------
+
+The types supported by UDFs are currently limited to:
+
+- int, Integer
+- boolean, Boolean
+- long, Long
+- double, Double
+- String
+- java.util.Map (must be a map of supported types)
+- java.util.List (must be a list of supported types)
+
+Note: Complex types are not currently supported
+
+
+=========
+Deploying
+=========
+
+To deploy your UDFs you need to create a jar containing all of the classes required by the UDFs.
+If you depend on third-party libraries then this should be an uber-jar containing those libraries.
+Once the jar is created you need to deploy it to each KSQL server instance. The jar should be copied
+to the ``ext`` directory that is part of the KSQL distribution.
+
+The jars in the ``ext`` directory are only scanned at start-up, so you will need to restart your
+KSQL server instances.
+
+
+=====
+Usage
+=====
+
+Once your UDFs are deployed you can call them in the same way you would invoke any of the KSQL
+builtin functions, for example, using the ``multiply`` example above:
+
+.. code:: sql
+
+    CREATE STREAM number_stream (int1 INT, int2 INT, long1 BIGINT, long2 BIGINT)
+      WITH (VALUE_FORMAT = 'JSON', KAFKA_TOPIC = 'numbers');
+
+    SELECT multiply(int1, int2), multiply(long1, long2) FROM number_stream;
+
+
+
+============
+Blacklisting
+============
+
+You can blacklist classes and packages such that they can't be used from a UDF. There is small
+blacklist that is found in the file ``resource-blacklist.txt`` that is in the ``ext`` directory.
+This file contains an entry per line, where each line is a class or package that should be blacklisted.
+The matching of the names is based on a regular expression, so if you have an entry, ``java.lang.Process``
+
+.. code:: txt
+
+    java.lang.Process
+
+This would match any paths that begin with java.lang.Process, i.e., java.lang.Process, java.lang.ProcessBuilder etc.
+
+Any blank lines or lines beginning with ``#`` are ignored. If the file is not present then all classes
+are blacklisted.
+
+
+Metric Collection
+=================
+
+TODO
+
+
+=========
+Disabling
+=========
+
+You can disable the loading of all UDFs in the ``ext`` directory by setting ``ksql.udfs.enabled`` to
```

Comment:
Question: This is the PR for UDFs, but we are also looking into UDAFs, for example.  Would we only have the one `ksql.udfs.enabled` config option, or separate ones for UDAFs eventually?

Reply:
I don't think the present work allows externally loaded UDAFs. Whenever we extend the framework to support loading UDAFs, I think they should also be considered UDFs, and will not need a different config.

Whether a given function is a simple transformation function or an aggregation function is an internal detail that doesn't need to bubble up to the top level, IMO.

Reply:
Yep, as @apurvam says